srtcore/epoll.cpp

```diff
@@ -569,12 +523,33 @@ int CEPoll::update_events(const SRTSOCKET& uid, std::set<int>& eids, int events,
       }
       else
       {
-         if ((events & UDT_EPOLL_IN) != 0)
-            update_epoll_sets(uid, p->second.m_sUDTSocksIn, p->second.m_sUDTReads, enable);
-         if ((events & UDT_EPOLL_OUT) != 0)
-            update_epoll_sets(uid, p->second.m_sUDTSocksOut, p->second.m_sUDTWrites, enable);
-         if ((events & UDT_EPOLL_ERR) != 0)
-            update_epoll_sets(uid, p->second.m_sUDTSocksEx, p->second.m_sUDTExcepts, enable);
+         const map<SRTSOCKET, pair<int, int>>::iterator& it = p->second.m_sUDTSocksWait.find(uid);
+         if(it==p->second.m_sUDTSocksWait.end())
+            continue;
+         // compute new states
+         pair<int, int>& wait(it->second);
+         if(enable)
+            events = wait.second | events;
+         else
+            events = wait.second & (~events);
+         // compute states changes!
+         int changes = wait.second ^ events; // oldState XOR newState
+         if(!changes)
+            continue; // no changes!
+         // assign new state
+         wait.second = events;
+         // filter change relating what is watching
+         if(!(changes &= wait.first))
+            continue; // no change watching
+         // set events changes!
+         if(!enable)
+         { // bits change from 1 to 0
+            pair<map<SRTSOCKET, int>::iterator, bool> it = p->second.m_sUDTSocksSet.insert(pair<SRTSOCKET, int>(uid, 0));
```

Comment:
Why can't it instead search for the record in the *Set map and then modify or erase it if needed?

Reply:
find+insert generate two dichotomous search rather only one, this algo is more efficient